connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java

```diff
@@ -288,6 +302,7 @@ private boolean sendRecords() {
         recordBatch(toSend.size());
         final SourceRecordWriteCounter counter = new SourceRecordWriteCounter(toSend.size(), sourceTaskMetricsGroup);
         for (final SourceRecord preTransformRecord : toSend) {
+            maybeThrowProducerSendException();
```

Comment:
do we need this? looks like the previous call is sufficient for most cases and is always guaranteed to be called. the only additional benefit here is that we would skip producing records.

Reply:
> the only additional benefit here is that we would skip producing records

I think that's a pretty large benefit and warrants the extra line here. Happy to remove if others disagree, though.

Reply:
If the producer failed to send a record, don't we want to stop immediately? Otherwise, we'd end up producing potentially out of order. The producer is set to allow at most only 1 in-flight record at a time, so it seems like we'd want to void producing as soon as we do this.

I guess the question is: will the producer ever fail on one message with a non-retriable exception, and then be good enough to continue writing any other records. If not, then we don't need this check. However, if that's a possibility, doesn't having this check here help us prevent writing a record after another record failed to send?

Reply:
> will the producer ever fail on one message with a non-retriable exception, and then be good enough to continue writing any other records

This is definitely a possibility. One example where this might come up is if two consecutive records are sent to two different topics; if the producer lacks permission to write to the first topic but has permission to write to the second, then the second will be sent successfully even though the first failed. This wouldn't be too big of an issue as long as the source partitions mapped nicely to Kafka topic partitions since that would ensure that the first record's source offset wouldn't be committed and would be re-sent as soon as the connector gets restarted, but since there's no enforced contract for those mappings I'd say it's not fair to make that assumption.